modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java

```diff
@@ -17,51 +17,310 @@
+     * </ol>
+     * So that, in case of initial node start (first start ever) specified configuration, supplemented with defaults, is
+     * used. If no configuration was provided defaults are used for all configuration properties. In case of node
+     * restart, specified properties override existing ones, non specified properties that also weren't specified
+     * previously use default values. Please pay attention that previously specified properties are searched in the
+     * {@code workDir} specified by the user.
+     */
+    public void start(@Nullable String cfg) {
+        List<IgniteComponent> startedComponents = new ArrayList<>();
+
+        try {
+            // Vault startup.
+            doStartComponent(
+                name,
+                startedComponents,
+                vaultMgr
+            );
+
+            vaultMgr.putName(name).join();
+
+            // Node configuration manager startup.
+            doStartComponent(
+                name,
+                startedComponents,
+                nodeCfgMgr);
+
+            // Node configuration manager bootstrap.
+            if (cfg != null) {
+                try {
+                    nodeCfgMgr.bootstrap(cfg);
+                }
+                catch (Exception e) {
+                    LOG.warn("Unable to parse user-specific configuration, default configuration will be used: {}",
+                        e.getMessage());
+                }
+            }
+            else
+                nodeCfgMgr.configurationRegistry().initializeDefaults();
+
+            // Start the remaining components.
+            List<IgniteComponent> otherComponents = List.of(
+                clusterSvc,
+                raftMgr,
+                metaStorageMgr,
+                clusterCfgMgr,
+                baselineMgr,
+                affinityMgr,
+                schemaMgr,
+                distributedTblMgr,
+                qryEngine,
+                restModule,
+                clientHandlerModule
+            );
+
+            for (IgniteComponent component : otherComponents)
+                doStartComponent(name, startedComponents, component);
+
+            // Deploy all resisted watches cause all components are ready and have registered their listeners.
+            metaStorageMgr.deployWatches();
+
+            if (!status.compareAndSet(Status.STARTING, Status.STARTED))
+                throw new NodeStoppingException();
+        }
+        catch (Exception e) {
+            String errMsg = "Unable to start node=[" + name + "].";
+
+            LOG.error(errMsg, e);
+
+            doStopNode(startedComponents);
+
+            throw new IgniteException(errMsg, e);
+        }
+    }
+
+    /**
+     * Stops ignite node.
+     */
+    public void stop() {
+        AtomicBoolean explicitStop = new AtomicBoolean();
+
+        status.getAndUpdate(status -> {
+            if (status == Status.STARTED)
+                explicitStop.set(true);
```

Comment:
I think there should be an `else` clause with `explicitStop.set(false)`, because this closure could be called multiple times (if stop is called from different threads) and `doStopNode` will be called multiple times

Reply:
Fixed.